File: geompp_tests/src/test_bprism3d.cpp
#include "bprism3d.hpp"
#include "point3d.hpp"
#include "utils.hpp"
#include "vector3d.hpp"

#include <gtest/gtest.h>

#include <cmath>

namespace g = geompp;

namespace geompp_tests {

class BPrism3DTest : public ::testing::Test {
 protected:
  void SetUp() override { g::DECIMAL_PRECISION = g::DP_THREE; }
  void TearDown() override { g::DECIMAL_PRECISION = g::DP_THREE; }
};

// ── Constructor ──────────────────────────────────────────────────────────────

TEST_F(BPrism3DTest, ConstructorEmpty_Throws) {
  ASSERT_THROW(g::BPrism3D(std::vector<g::Point3D>{}), std::invalid_argument);
}

TEST_F(BPrism3DTest, ConstructorSinglePoint_Throws) {
  ASSERT_THROW(g::BPrism3D(std::vector<g::Point3D>{{3.0, 4.0, 5.0}}), std::invalid_argument);
}

TEST_F(BPrism3DTest, ConstructorTwoPoints_Throws) {
  ASSERT_THROW(g::BPrism3D(std::vector<g::Point3D>{{0.0, 0.0, 0.0}, {4.0, 0.0, 0.0}}),
               std::invalid_argument);
}

TEST_F(BPrism3DTest, ConstructorAxisAlignedBox) {
  // 8 corners of a 4x3x2 box
  std::vector<g::Point3D> pts = {
    {0, 0, 0}, {4, 0, 0}, {4, 3, 0}, {0, 3, 0},
    {0, 0, 2}, {4, 0, 2}, {4, 3, 2}, {0, 3, 2},
  };
  auto p = g::BPrism3D(pts);
  ASSERT_TRUE(p.center().AlmostEquals(g::Point3D(2.0, 1.5, 1.0)));
  ASSERT_NEAR(p.volume(), 24.0, 1e-6);
  // All input points must be contained
  for (auto const& pt : pts) {
    ASSERT_TRUE(p.Contains(pt)) << "prism must contain all input points";
  }
}

TEST_F(BPrism3DTest, ConstructorFlatCloud_WIsEpsilon) {
  // 4 coplanar points in the z=0 plane — HALF_LEN_W should be clamped to DOUBLE_EPSILON
  std::vector<g::Point3D> pts = {
    {0, 0, 0}, {2, 0, 0}, {2, 1, 0}, {0, 1, 0},
  };
  auto p = g::BPrism3D(pts);
  double eps = g::DOUBLE_EPSILON;
  ASSERT_NEAR(p.half_len_w(), eps, 1e-12);
}

TEST_F(BPrism3DTest, ConstructorNonConvex_AllPointsContained) {
  // Box corners plus interior points — every input point must be inside
  std::vector<g::Point3D> pts = {
    {0, 0, 0}, {4, 0, 0}, {4, 3, 0}, {0, 3, 0},
    {0, 0, 2}, {4, 0, 2}, {4, 3, 2}, {0, 3, 2},
    {1, 1, 1}, {2, 0.5, 0.3}, {3.5, 2.5, 1.8},
  };
  auto p = g::BPrism3D(pts);
  for (auto const& pt : pts) {
    ASSERT_TRUE(p.Contains(pt)) << "prism must contain all input points";
  }
}

// ── Accessors ────────────────────────────────────────────────────────────────

TEST_F(BPrism3DTest, Accessors_AxesAreUnitVectors) {
  std::vector<g::Point3D> pts = {
    {0, 0, 0}, {4, 0, 0}, {4, 3, 0}, {0, 3, 0},
    {0, 0, 2}, {4, 0, 2}, {4, 3, 2}, {0, 3, 2},
  };
  auto p = g::BPrism3D(pts);
  ASSERT_NEAR(p.axis_u().Length(), 1.0, 1e-9);
  ASSERT_NEAR(p.axis_v().Length(), 1.0, 1e-9);
  ASSERT_NEAR(p.axis_w().Length(), 1.0, 1e-9);
}

TEST_F(BPrism3DTest, Accessors_AxesOrthogonal) {
  std::vector<g::Point3D> pts = {
    {0, 0, 0}, {4, 0, 0}, {4, 3, 0}, {0, 3, 0},
    {0, 0, 2}, {4, 0, 2}, {4, 3, 2}, {0, 3, 2},
  };
  auto p = g::BPrism3D(pts);
  ASSERT_NEAR(p.axis_u().Dot(p.axis_v()), 0.0, 1e-9);
  ASSERT_NEAR(p.axis_u().Dot(p.axis_w()), 0.0, 1e-9);
  ASSERT_NEAR(p.axis_v().Dot(p.axis_w()), 0.0, 1e-9);
}

TEST_F(BPrism3DTest, Accessors_WidthHeightDepthVolume) {
  std::vector<g::Point3D> pts = {
    {0, 0, 0}, {4, 0, 0}, {4, 3, 0}, {0, 3, 0},
    {0, 0, 2}, {4, 0, 2}, {4, 3, 2}, {0, 3, 2},
  };
  auto p = g::BPrism3D(pts);
  ASSERT_NEAR(p.width(), 2.0 * p.half_len_u(), 1e-9);
  ASSERT_NEAR(p.height(), 2.0 * p.half_len_v(), 1e-9);
  ASSERT_NEAR(p.depth(), 2.0 * p.half_len_w(), 1e-9);
  ASSERT_NEAR(p.volume(), p.width() * p.height() * p.depth(), 1e-9);
}

// ── Corners ──────────────────────────────────────────────────────────────────

TEST_F(BPrism3DTest, Corners_EightDistinctPoints) {
  std::vector<g::Point3D> pts = {
    {0, 0, 0}, {4, 0, 0}, {4, 3, 0}, {0, 3, 0},
    {0, 0, 2}, {4, 0, 2}, {4, 3, 2}, {0, 3, 2},
  };
  auto p = g::BPrism3D(pts);
  auto corners = p.Corners();
  ASSERT_EQ(corners.size(), 8u);
  // All eight corners must be contained in the prism
  for (auto const& c : corners) {
    ASSERT_TRUE(p.Contains(c)) << "prism must contain all its corners";
  }
}

// ── Contains ─────────────────────────────────────────────────────────────────

TEST_F(BPrism3DTest, Contains_Center_True) {
  g::DECIMAL_PRECISION = 6;
  std::vector<g::Point3D> pts = {
    {0, 0, 0}, {4, 0, 0}, {4, 3, 0}, {0, 3, 0},
    {0, 0, 2}, {4, 0, 2}, {4, 3, 2}, {0, 3, 2},
  };
  auto p = g::BPrism3D(pts);
  ASSERT_TRUE(p.Contains(p.center()));
}

TEST_F(BPrism3DTest, Contains_Interior_True) {
  g::DECIMAL_PRECISION = 6;
  std::vector<g::Point3D> pts = {
    {0, 0, 0}, {4, 0, 0}, {4, 3, 0}, {0, 3, 0},
    {0, 0, 2}, {4, 0, 2}, {4, 3, 2}, {0, 3, 2},
  };
  auto p = g::BPrism3D(pts);
  ASSERT_TRUE(p.Contains(g::Point3D(1.0, 1.0, 0.5)));
  ASSERT_TRUE(p.Contains(g::Point3D(3.5, 2.5, 1.8)));
}

TEST_F(BPrism3DTest, Contains_Outside_False) {
  g::DECIMAL_PRECISION = 6;
  std::vector<g::Point3D> pts = {
    {0, 0, 0}, {4, 0, 0}, {4, 3, 0}, {0, 3, 0},
    {0, 0, 2}, {4, 0, 2}, {4, 3, 2}, {0, 3, 2},
  };
  auto p = g::BPrism3D(pts);
  ASSERT_FALSE(p.Contains(g::Point3D(5.0, 1.5, 1.0)));   // beyond u-axis
  ASSERT_FALSE(p.Contains(g::Point3D(2.0, 4.0, 1.0)));   // beyond v-axis
  ASSERT_FALSE(p.Contains(g::Point3D(2.0, 1.5, 3.0)));   // beyond w-axis
}

TEST_F(BPrism3DTest, Contains_Boundary_True) {
  g::DECIMAL_PRECISION = 6;
  std::vector<g::Point3D> pts = {
    {0, 0, 0}, {4, 0, 0}, {4, 3, 0}, {0, 3, 0},
    {0, 0, 2}, {4, 0, 2}, {4, 3, 2}, {0, 3, 2},
  };
  auto p = g::BPrism3D(pts);
  // Input corners lie on the prism boundary
  ASSERT_TRUE(p.Contains(g::Point3D(0.0, 0.0, 0.0)));
  ASSERT_TRUE(p.Contains(g::Point3D(4.0, 3.0, 2.0)));
}

// ── AlmostEquals / operator== ─────────────────────────────────────────────────

TEST_F(BPrism3DTest, AlmostEquals_Same) {
  std::vector<g::Point3D> pts = {
    {0, 0, 0}, {4, 0, 0}, {4, 3, 0}, {0, 3, 0},
    {0, 0, 2}, {4, 0, 2}, {4, 3, 2}, {0, 3, 2},
  };
  auto p1 = g::BPrism3D(pts);
  auto p2 = g::BPrism3D(pts);
  ASSERT_TRUE(p1.AlmostEquals(p2));
  ASSERT_EQ(p1, p2);
}

TEST_F(BPrism3DTest, AlmostEquals_Different) {
  std::vector<g::Point3D> pts1 = {
    {0, 0, 0}, {4, 0, 0}, {4, 3, 0}, {0, 3, 0},
    {0, 0, 2}, {4, 0, 2}, {4, 3, 2}, {0, 3, 2},
  };
  std::vector<g::Point3D> pts2 = {
    {0, 0, 0}, {6, 0, 0}, {6, 3, 0}, {0, 3, 0},
    {0, 0, 2}, {6, 0, 2}, {6, 3, 2}, {0, 3, 2},
  };
  auto p1 = g::BPrism3D(pts1);
  auto p2 = g::BPrism3D(pts2);
  ASSERT_FALSE(p1.AlmostEquals(p2));
  ASSERT_NE(p1, p2);
}

// ── Copy / Assignment ─────────────────────────────────────────────────────────

TEST_F(BPrism3DTest, CopyConstructor) {
  std::vector<g::Point3D> pts = {
    {0, 0, 0}, {4, 0, 0}, {4, 3, 0}, {0, 3, 0},
    {0, 0, 2}, {4, 0, 2}, {4, 3, 2}, {0, 3, 2},
  };
  auto p1 = g::BPrism3D(pts);
  auto p2 = p1;
  ASSERT_TRUE(p1.AlmostEquals(p2));
}

TEST_F(BPrism3DTest, Assignment) {
  std::vector<g::Point3D> pts1 = {
    {0, 0, 0}, {4, 0, 0}, {4, 3, 0}, {0, 3, 0},
    {0, 0, 2}, {4, 0, 2}, {4, 3, 2}, {0, 3, 2},
  };
  std::vector<g::Point3D> pts2 = {
    {0, 0, 0}, {1, 0, 0}, {1, 1, 0}, {0, 1, 0},
    {0, 0, 1}, {1, 0, 1}, {1, 1, 1}, {0, 1, 1},
  };
  auto p1 = g::BPrism3D(pts1);
  auto p2 = g::BPrism3D(pts2);
  p2 = p1;
  ASSERT_TRUE(p2.AlmostEquals(p1));
  p1 = p1;  // self-assignment
  ASSERT_TRUE(p1.AlmostEquals(g::BPrism3D(pts1)));
}

}  // namespace geompp_tests

File: geompp_python/tests/test_bprism3d.py
"""
bprism3d binding tests.
"""

import math
import os
import tempfile
import pytest
import geompp

from ._helpers import approx


class TestBPrism3D:
    def test_empty_throws(self):
        with pytest.raises(Exception):
            geompp.BPrism3D([])

    def test_single_point_throws(self):
        with pytest.raises(Exception):
            geompp.BPrism3D([geompp.Point3D(3, 4, 5)])

    def test_two_points_throws(self):
        with pytest.raises(Exception):
            geompp.BPrism3D([geompp.Point3D(0, 0, 0), geompp.Point3D(4, 0, 0)])

    def test_axis_aligned_box(self):
        pts = [
            geompp.Point3D(0, 0, 0), geompp.Point3D(4, 0, 0),
            geompp.Point3D(4, 3, 0), geompp.Point3D(0, 3, 0),
            geompp.Point3D(0, 0, 2), geompp.Point3D(4, 0, 2),
            geompp.Point3D(4, 3, 2), geompp.Point3D(0, 3, 2),
        ]
        p = geompp.BPrism3D(pts)
        assert approx(p.volume, 24.0, eps=0.5)
        for pt in pts:
            assert p.contains(pt), f"prism must contain input point {pt}"

    def test_flat_cloud_w_is_epsilon(self):
        pts = [
            geompp.Point3D(0, 0, 0), geompp.Point3D(4, 0, 0),
            geompp.Point3D(4, 3, 0), geompp.Point3D(0, 3, 0),
        ]
        p = geompp.BPrism3D(pts)
        assert p.half_len_w > 0, "flat cloud must have non-zero w half-length"
        assert approx(p.half_len_w, 0.001, eps=0.01)

    def test_nonconvex_all_points_contained(self):
        pts = [
            geompp.Point3D(0, 0, 0), geompp.Point3D(3, 0, 0),
            geompp.Point3D(3, 2, 0), geompp.Point3D(0, 2, 0),
            geompp.Point3D(1, 0.5, 1), geompp.Point3D(2, 1.5, 0.5),
        ]
        p = geompp.BPrism3D(pts)
        for pt in pts:
            assert p.contains(pt), f"prism must contain input point {pt}"

    def test_axes_are_unit_vectors(self):
        pts = [
            geompp.Point3D(0, 0, 0), geompp.Point3D(4, 0, 0),
            geompp.Point3D(4, 3, 0), geompp.Point3D(0, 3, 0),
            geompp.Point3D(0, 0, 2), geompp.Point3D(4, 0, 2),
            geompp.Point3D(4, 3, 2), geompp.Point3D(0, 3, 2),
        ]
        p = geompp.BPrism3D(pts)
        assert approx(p.axis_u.length(), 1.0)
        assert approx(p.axis_v.length(), 1.0)
        assert approx(p.axis_w.length(), 1.0)

    def test_axes_orthogonal(self):
        pts = [
            geompp.Point3D(0, 0, 0), geompp.Point3D(4, 0, 0),
            geompp.Point3D(4, 3, 0), geompp.Point3D(0, 3, 0),
            geompp.Point3D(0, 0, 2), geompp.Point3D(4, 0, 2),
            geompp.Point3D(4, 3, 2), geompp.Point3D(0, 3, 2),
        ]
        p = geompp.BPrism3D(pts)
        assert approx(p.axis_u.dot(p.axis_v), 0.0)
        assert approx(p.axis_u.dot(p.axis_w), 0.0)
        assert approx(p.axis_v.dot(p.axis_w), 0.0)

    def test_width_height_depth_volume(self):
        pts = [
            geompp.Point3D(0, 0, 0), geompp.Point3D(4, 0, 0),
            geompp.Point3D(4, 3, 0), geompp.Point3D(0, 3, 0),
            geompp.Point3D(0, 0, 2), geompp.Point3D(4, 0, 2),
            geompp.Point3D(4, 3, 2), geompp.Point3D(0, 3, 2),
        ]
        p = geompp.BPrism3D(pts)
        assert approx(p.width, 2.0 * p.half_len_u)
        assert approx(p.height, 2.0 * p.half_len_v)
        assert approx(p.depth, 2.0 * p.half_len_w)
        assert approx(p.volume, p.width * p.height * p.depth, eps=0.5)

    def test_corners_eight_distinct_all_contained(self):
        pts = [
            geompp.Point3D(0, 0, 0), geompp.Point3D(4, 0, 0),
            geompp.Point3D(4, 3, 0), geompp.Point3D(0, 3, 0),
            geompp.Point3D(0, 0, 2), geompp.Point3D(4, 0, 2),
            geompp.Point3D(4, 3, 2), geompp.Point3D(0, 3, 2),
        ]
        p = geompp.BPrism3D(pts)
        corners = p.corners()
        assert len(corners) == 8
        for c in corners:
            assert p.contains(c), "each corner must be inside the prism"

    def test_contains_center(self):
        pts = [
            geompp.Point3D(0, 0, 0), geompp.Point3D(4, 0, 0),
            geompp.Point3D(4, 3, 0), geompp.Point3D(0, 3, 0),
            geompp.Point3D(0, 0, 2), geompp.Point3D(4, 0, 2),
            geompp.Point3D(4, 3, 2), geompp.Point3D(0, 3, 2),
        ]
        p = geompp.BPrism3D(pts)
        assert p.contains(p.center)

    def test_contains_outside_false(self):
        pts = [
            geompp.Point3D(0, 0, 0), geompp.Point3D(4, 0, 0),
            geompp.Point3D(4, 3, 0), geompp.Point3D(0, 3, 0),
            geompp.Point3D(0, 0, 2), geompp.Point3D(4, 0, 2),
            geompp.Point3D(4, 3, 2), geompp.Point3D(0, 3, 2),
        ]
        p = geompp.BPrism3D(pts)
        assert not p.contains(geompp.Point3D(10, 10, 10))

    def test_almost_equals(self):
        pts = [
            geompp.Point3D(0, 0, 0), geompp.Point3D(4, 0, 0),
            geompp.Point3D(4, 3, 0), geompp.Point3D(0, 3, 0),
            geompp.Point3D(0, 0, 2), geompp.Point3D(4, 0, 2),
            geompp.Point3D(4, 3, 2), geompp.Point3D(0, 3, 2),
        ]
        p1 = geompp.BPrism3D(pts)
        p2 = geompp.BPrism3D(pts)
        assert p1.almost_equals(p2)
        assert p1 == p2

    def test_not_almost_equals_different(self):
        pts1 = [
            geompp.Point3D(0, 0, 0), geompp.Point3D(4, 0, 0),
            geompp.Point3D(4, 3, 0), geompp.Point3D(0, 3, 0),
        ]
        pts2 = [
            geompp.Point3D(0, 0, 0), geompp.Point3D(6, 0, 0),
            geompp.Point3D(6, 3, 0), geompp.Point3D(0, 3, 0),
        ]
        p1 = geompp.BPrism3D(pts1)
        p2 = geompp.BPrism3D(pts2)
        assert not p1.almost_equals(p2)
